src/utilities/BatchTableHierarchyExtension.js

```diff
@@ -0,0 +1,123 @@
+import { parseBinArray } from './FeatureTable.js';
+
+export class BatchTableHierarchyExtension {
+
+	constructor( batchTable ) {
+
+		this.batchTable = batchTable;
+
+		const extensionHeader = batchTable.header.extensions[ '3DTILES_batch_table_hierarchy' ];
+
+		this.classes = extensionHeader.classes;
+		for ( const classDef of this.classes ) {
+
+			const instances = classDef.instances;
+			for ( const property in instances ) {
+
+				classDef.instances[ property ] = this._parseProperty( instances[ property ], classDef.length );
+
+			}
+
+		}
+
+		this.instancesLength = extensionHeader.instancesLength;
+
+		this.classIds = this._parseProperty( extensionHeader.classIds, this.instancesLength );
+		this.parentCounts = this._parseProperty( extensionHeader.parentCounts, this.instancesLength )
+			|| new Array( this.instancesLength ).fill( 1 );
+		this.parentIds = this._parseProperty( extensionHeader.parentIds );
+
+		this.instancesIds = [];
+		const classCounter = {};
+		for ( const classId of this.classIds ) {
+
+			classCounter[ classId ] = classCounter[ classId ] ?? 0;
+			this.instancesIds.push( classCounter[ classId ] );
+			classCounter[ classId ] ++;
+
+		}
+
+	}
+
+	_parseProperty( property, propertyLength ) {
+
+		if ( ! ( property instanceof Object ) ) {
+
+			return property;
+
+		} else if ( Array.isArray( property ) ) {
+
+			return property;
+
+		} else {
+
+			const { buffer, binOffset } = this.batchTable;
+
+			const byteOffset = property.byteOffset;
+			const componentType = property.componentType || 'UNSIGNED_SHORT';
+
+			const arrayStart = binOffset + byteOffset;
+
+			return parseBinArray( buffer, arrayStart, propertyLength, 'SCALAR', componentType );
```

Comment:
The only way an error that uses the property name can be thrown when `parseBinArray` is called from `BatchTableHierarchyExtension` is having a `componentType` in data that is neither `undefined` nor one of the types described [here](https://github.com/CesiumGS/3d-tiles/blob/87218a7c2030890b5200cf16cb466dbbd1480a20/extensions/3DTILES_batch_table_hierarchy/schema/3DTILES_batch_table_hierarchy.schema.json#L56-L86).

Currently in `BatchTableHierarchyExtension`, an `undefined` `componentType` defaults to `'UNSIGNED_SHORT'` and `type` is hard coded as `'SCALAR'` (following the specification). Therefore, I think it is quite unlikely that the error messages thrown in `parseBinArray` is ever thrown, except for some weirdly ill-formatted data.

Moreover `BatchTableHierarchyExtension._parseProperty` is parametrized with property values instead of name (which is not the case in `FeatureTable.getData`, the other place `parseBinArray` is called from). I made it that way so that any property can be parsed using `_parseProperty`, and not only properties that share the same parent object. We can thus parse `extensionHeader.classIds` with the same method we use to parse `extensionHeader.classes[ x ].instances[ someProperty ]`). We would therefore need to add a parameter to `BatchTableHierarchyExtension._parseProperty` just to pass the property name (which would be hard coded each time).

Therefore, I would suggest not to add the property name in this situation, though I'll add it if you still want it added :smile:

Reply:
Got it - we should add something for an argument here or set the `propertyName` field to use some kind of reasonable default as a name in the function parameters, then. At the moment it looks like a bug, I feel.

Reply:
Okay, I'll add the property name as an argument then.